turbopack/crates/turbo-tasks/src/scope.rs

```diff
@@ -2,33 +2,43 @@
 
 use std::{
     any::Any,
+    collections::VecDeque,
     marker::PhantomData,
     panic::{self, AssertUnwindSafe, catch_unwind},
-    pin::Pin,
     sync::{
         Arc,
         atomic::{AtomicUsize, Ordering},
     },
-    thread::{self, Thread},
+    thread::{self, Thread, available_parallelism},
     time::{Duration, Instant},
 };
 
-use futures::FutureExt;
-use parking_lot::Mutex;
+use once_cell::sync::Lazy;
+use parking_lot::{Condvar, Mutex};
 use tokio::{runtime::Handle, task::block_in_place};
-use tracing::{Instrument, Span, info_span};
+use tracing::{Span, info_span};
 
-use crate::{
-    TurboTasksApi,
-    manager::{try_turbo_tasks, turbo_tasks_future_scope},
-};
+use crate::{TurboTasksApi, manager::try_turbo_tasks, turbo_tasks_scope};
+
+/// Number of worker tasks to spawn that process jobs. It's 1 less than the number of cpus as we
+/// also use the current task as worker.
+static WORKER_TASKS: Lazy<usize> = Lazy::new(|| available_parallelism().map_or(0, |n| n.get() - 1));
+
+enum WorkQueueJob {
+    Job(usize, Box<dyn FnOnce() + Send + 'static>),
+    End,
```

Comment:
oh i think i get it

this is a 'poison pill' for the queue to get workers to stop.  this is why workers add it back after reading it, to ensure all workers are stopped.

add a doc

it also isn't obvious to me what the `usize` represents in the `Job` so use named struct fields?